### DISCUSSION TITLE
DB override in unit tests makes client don't work as expected

### DISCUSSION THREAD

**Discussion by lays147** [fastapi/fastapi#6878] [Questions] [answered]
### Problem
I'm writing an application that uses SQLAlchemy as a database. Most of the setup is based on FastAPI docs.
When I'm unit testing the routers, I needed to update the ``app`` to override the ``get_db`` method. Apparently, the client doesn't build correctly and returns ``422`` as an error, the function related doesn't even start to process.
Any ideas on how I can debug this?

``main.py``
```python
@pytest.fixture
def client(db) -> TestClient:
    def _overwrite_db(db):
        return db
    app.dependency_overrides[get_db] = _overwrite_db
    return TestClient(app)
```

The DB fixtures are defined by this gist: https://gist.github.com/kissgyorgy/e2365f25a213de44b9a2

```python
    def test_partner_already_exists(client: TestClient, valid_body: dict) -> None:
        response = client.post('/partners', json=valid_body)
>       assert response.status_code == status.HTTP_409_CONFLICT, 'Status code != 409'
E       AssertionError: Status code != 409
E       assert 422 == 409
E        +  where 422 = <Response [422]>.status_code
E        +  and   409 = status.HTTP_409_CONFLICT
```

## System
Python 3.8.5

click==7.1.2
python-dotenv==0.15.0
fastapi==0.63.0
h11==0.12.0
psycopg2==2.8.6
pydantic==1.7.3
shapely==1.7.1
SQLAlchemy==1.3.23
starlette==0.13.6
uvicorn==0.13.4

**Comment by awtkns** [comment]
I think your `valid_body` dict does not conform to the model you set up in your route. An http 422 is normally a pydantic validation error. 

You might be able to find more about the validation error itself by logging `response.json()` in your test.

**Comment by Kludex** [comment, marked as answer]
Instead of:

```python
assert​ ​response​.​status_code​ ​==​ ​status.HTTP_409_CONFLICT, ​'Status code != 409'
```
Wouldn't it be better if you:

```python
assert​ ​response​.​status_code​ ​==​ ​status.HTTP_409_CONFLICT, ​res.json()
```

Anyway, the issue is probably the parameter you have in the new dependency. Solution would be (possibly, if I'm not mistaken):
```python
    ​def​ ​_overwrite_db​(​):
        ​return​ ​db
```

**Comment by lays147** [comment]
Hi @Kludex, thanks for the reply.  The removal of the parameter in the inside function solved my issue.

**Comment by tiangolo** [comment]
Thanks for the help here @Kludex and @awtkns ! 👏  🙇 

Thanks for reporting back and closing the issue @lays147  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
